src/main/java/uk/gov/pay/api/filter/ratelimit/RedisRateLimiter.java

```diff
@@ -0,0 +1,103 @@
+package uk.gov.pay.api.filter.ratelimit;
+
+import com.google.inject.Inject;
+import com.google.inject.OutOfScopeException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPool;
+
+import java.time.LocalDateTime;
+import java.time.temporal.ChronoField;
+
+public class RedisRateLimiter {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RedisRateLimiter.class);
+
+    private JedisPool jedisPool;
+
+    @Inject
+    public RedisRateLimiter(JedisPool jedisPool) {
+        this.jedisPool = jedisPool;
+    }
+
+    /**
+     * This block needs to be synchronous. Each RateLimit object will be shared between requests
+     * from the same source (Service), so is not shared across all the requests.
+     *
+     * @throws RateLimitException
+     */
+    synchronized void checkRateOf(String key, int rate, int perMillis)
+            throws RedisRateLimitException, RateLimitException {
+
+        Long count = null;
+
+        try {
+            count=updateAllowance(key, perMillis);
+        }
+        catch(Exception e){
+            // Exception possible if redis is unavailable or perMillis is too low or too high        
+            throw new RedisRateLimitException();
+        }
+
+        if (count!=null && count > rate) {
+            LOGGER.info(String.format("rate exceeded - count: %d, rate: %d", count, rate));
+            throw new RateLimitException();
+        }
+    }
+
+    synchronized private Long updateAllowance(String key, int perMillis) {
+
+        String derivedKey = getKeyForWindow(key, perMillis);
+        Long count = getResource().incr(derivedKey);
+
+        if (count == 1) {
+            getResource().expire(derivedKey, perMillis/1000);
+        }
+
+        return count;
+    }
+
+    private Jedis getResource() {
+        return jedisPool.getResource();
+    }
+
+    /**
+     * Derives Key (Service Key + Window) to use in Redis for rate limiting.
+     * Recommended to use perMillis to lowest granularity. i.e, to seconds. 1000, 2000
+     * <p>
+     * Depends on perMillis
+     * <p>
+     * - perMillis < 1000 : Window calculated for milliseconds    
+     * - perMillis >=1000 && <=60000 : Window calculated for seconds(s)
+     * - perMillis >60000 && <=3600000 : Window calculated for minute(s)
+     * - perMillis >3600000 && <=86400000 : Window calculated for hours(s)
+     *
+     * @param perMillis
+     * @return new key based on perMillis (works for second/minute/hour windows only)
+     */
+    private String getKeyForWindow(String key, int perMillis) throws OutOfScopeException {
+
+        LocalDateTime now = LocalDateTime.now();
+
+        int slot;
```

Comment:
rename `slot` to `window` for consistency, since you're talking about "Window" in the javadoc

Reply:
Renamed..